pkg/modprovider/module.go

```diff
@@ -0,0 +1,595 @@
+
+	return &pulumirpc.CreateResponse{
+		Id:         moduleStateResourceID,
+		Properties: props,
+	}, nil
+}
+
+func (h *moduleHandler) Update(
+	ctx context.Context,
+	req *pulumirpc.UpdateRequest,
+	moduleSource TFModuleSource,
+	moduleVersion TFModuleVersion,
+	providersConfig map[string]resource.PropertyMap,
+	inferredModule *InferredModuleSchema,
+	packageName packageName,
+) (*pulumirpc.UpdateResponse, error) {
+	urn := urn.URN(req.GetUrn())
+	logger := newResourceLogger(h.hc, urn)
+
+	moduleInputs, err := plugin.UnmarshalProperties(req.GetNews(), h.marshalOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	oldOutputs, err := plugin.UnmarshalProperties(req.GetOlds(), h.marshalOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	statusClient, err := h.statusPool.Acquire(ctx, logger, req.ResourceStatusAddress)
+	if err != nil {
+		return nil, err
+	}
+	defer statusClient.Release()
+
+	//q.Q("Update", req.GetPreview())
+
+	moduleOutputs, views, err := h.applyModuleOperation(
+		ctx,
+		urn,
+		moduleInputs,
+		oldOutputs,
+		moduleSource,
+		moduleVersion,
+		providersConfig,
+		inferredModule,
+		packageName,
+		req.GetPreview(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = statusClient.PublishViewSteps(ctx, &pulumirpc.PublishViewStepsRequest{
+		Token: req.ResourceStatusToken,
+		Steps: views,
+	})
+	if err != nil {
+		logger.Log(ctx, tfsandbox.Debug, fmt.Sprintf("error publishing view steps after Update: %v", err))
+		return nil, err
+	}
+
+	props, err := plugin.MarshalProperties(moduleOutputs, h.marshalOpts())
+	contract.AssertNoErrorf(err, "plugin.MarshalProperties should not fail")
+
+	return &pulumirpc.UpdateResponse{
+		Properties: props,
+	}, nil
+}
+
+// Delete calls TF Destroy to remove everything.
+func (h *moduleHandler) Delete(
+	ctx context.Context,
+	req *pulumirpc.DeleteRequest,
+	packageName packageName,
+	moduleSource TFModuleSource,
+	moduleVersion TFModuleVersion,
+	inferredModule *InferredModuleSchema,
+	providersConfig map[string]resource.PropertyMap,
+) (*emptypb.Empty, error) {
+	urn := urn.URN(req.GetUrn())
+	logger := newResourceLogger(h.hc, resource.URN(req.GetUrn()))
+
+	statusClient, err := h.statusPool.Acquire(ctx, logger, req.ResourceStatusAddress)
+	if err != nil {
+		return nil, err
+	}
+	defer statusClient.Release()
+
+	moduleInputs, err := plugin.UnmarshalProperties(req.GetOldInputs(), h.marshalOpts())
+	if err != nil {
+		return nil, fmt.Errorf("Delete failed to unmarshal inputs: %s", err)
+	}
+
+	oldOutputs, err := plugin.UnmarshalProperties(req.GetProperties(), h.marshalOpts())
+	if err != nil {
+		return nil, fmt.Errorf("Delete failed to unmarshal old outputs: %s", err)
+	}
+
+	tf, err := h.prepSandbox(
+		ctx,
+		urn,
+		moduleInputs,
+		oldOutputs,
+		inferredModule,
+		moduleSource,
+		moduleVersion,
+		providersConfig,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed preparing tofu sandbox: %w", err)
+	}
+
+	stateBeforeDestroy, err := tf.Show(ctx, logger)
```

Comment:
I'm not sure if this is an issue here or in the engine. When `delete` is run I only see the modules being deleted on the initial preview. Once the delete starts running and the views update I see the individual views being deleted. I think we should be able to see all the views on the initial delete preview.

```console
     Type                       Name               Plan       Info
 -   pulumi:pulumi:Stack        alb-example-chall  delete     9 warnings
 -   ├─ bucketmod:index:Module  test-log-bucket    delete
 -   ├─ lambdamod:index:Module  test-lambda        delete
 -   ├─ vpcmod:index:Module     test-vpc           delete
 -   └─ albmod:index:Module     test-alb           delete


```

Reply:
Hmm this doesn't look right, thanks for paying attention here! Let me debug this a little bit, either we're not sending the right view-steps or the engine is not rendering them yet as it should.

Reply:
Ok cross-linked follow-up work here.